spring-cloud-gcp-core/src/main/java/org/springframework/cloud/gcp/core/UsageTrackingHeaderProvider.java

```diff
@@ -0,0 +1,51 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.core;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.api.gax.rpc.HeaderProvider;
+
+/**
+ * Provides the User-Agent header to signal to the Google Cloud Client Libraries that requests originate from a Spring
+ * Integration.
+ *
+ * @author João André Martins
+ */
+public class UsageTrackingHeaderProvider implements HeaderProvider {
+
+	private final String springLibrary;
+
+	private final String springLibraryVersion;
+
+	public UsageTrackingHeaderProvider(String springLibrary, String springLibraryVersion) {
```

Comment:
I think you can replace `String springLibraryVersion` with just a `Class clazz`, and then use `this.clazz.getPackage().getImplementationVersion()` in `getHeaders`. This would make the users of this class a bit less verbose.

Furthermore, you can remove `springLibrary` argument altogether, and replace it with:
```java
		String[] pkg = getClass().getPackage().getName().split("\\.");
		String springLibrary = "spring-cloud-gcp-" + pkg[pkg.length-1];
```
You would then just end up with a constructor that looks like this:
`UsageTrackingHeaderProvider(Class clazz)`

Convention over configuration FTW!

Reply:
1) I want to be able to make it possible for the library version to be different from the project version. It might always be the same, but there are times we might need it to be different.

2) `getClass()` will always return `UsageTrackingHeaderProvider`, therefore `springLibrary` would always be `spring-cloud-gcp-core` IIUC?

Reply:
1. `clazz.getPackage().getImplementationVersion()` would return the version of the module in which the passed in class resides, right? So, it can be different from the top-level project version. Am I missing something? Currently, you're passing the same thing everywhere. If this changes some day, we can add another constructor then.

2. I meant to say `clazz.getClass()...` which would return the package name of the passed in class that doesn't have to be `UseTrackingHeaderProvider`.

Reply:
Aah, so you're suggesting passing in a class and then getting the data we need from it!
Yeah sure.

Reply:
One problem with this approach is that the classes calling the constructor are all in the `autoconfigure` package, so we would need to pass in another class in the feature's main module.
But some features might not even have their own modules (e.g., there were talks of getting rid of the config module).

I'm more inclined towards abandoning this idea, but wdyt?

Reply:
They don't have to be in their own modules, just in separate packages.
You're using `this.getClass()` everywhere currently. In what instance do you need to pass in a different class? In any case, it shouldn't pose an issue.

Reply:
@meltsufin I was passing classes in the `autoconfigure` module.
In the end, `springLibrary` ended up always being `spring-cloud-gcp-autoconfigure` which isn't helpful.

I don't think that suggestion is very feasible. I prefer being able to pass any value I want to report, instead of relying on a convention where the classes passed must be in a package whose last section is the service (e.g., `storage`, etc.).